gms/gossiper.cc

```diff
@@ -833,7 +838,38 @@ future<gossiper::endpoint_permit> gossiper::lock_endpoint(inet_address ep, permi
     }
     pid = permit_id::create_random_id();
     logger.debug("{}: lock_endpoint {}: waiting: permit_id={}", l.function_name(), ep, pid);
-    eptr->units = co_await get_units(eptr->sem, 1, _abort_source);
+    while (true) {
+        _abort_source.check();
+        static constexpr auto duration = std::chrono::minutes{1};
+        abort_on_expiry aoe(lowres_clock::now() + duration);
+        auto sub = _abort_source.subscribe([&aoe] () noexcept {
+            aoe.abort_source().request_abort();
+        });
+        assert(sub); // due to check() above
+        try {
+            eptr->units = co_await get_units(eptr->sem, 1, aoe.abort_source());
+            break;
+        } catch (const abort_requested_exception&) {
+            if (!_abort_source.abort_requested()) {
+                static constexpr auto fmt_loc = [] (const seastar::compat::source_location& l) {
+                    return fmt::format("{}({}:{}) `{}`", l.file_name(), l.line(), l.column(), l.function_name());
+                };
+                static constexpr auto fmt_loc_opt = [] (const std::optional<seastar::compat::source_location>& l) {
+                    if (!l) {
+                        return "null"s;
+                    }
+                    return fmt_loc(*l);
+                };
+                // The abort had to come from `abort_on_expiry`'s timer.
+                logger.error(
+                    "{}: waiting for endpoint lock (ep={}) took more than {}, signifying possible deadlock;"
+                    " holders: {}, first holder: {}, last holder (might not be current): {}",
+                    fmt_loc(l), ep, duration, eptr->holders, fmt_loc_opt(eptr->first_holder), fmt_loc_opt(eptr->last_holder));
+            } else {
+                throw;
+            }
```

Comment:
nit: I'd reverse the conditions to throw `if (_abort_source.abort_requested())` to reduce nesting.